hatchery/nextflow.go

```diff
@@ -838,11 +869,96 @@ func createS3bucket(s3Svc *s3.S3, bucketName string) error {
 		// no need to check for a specific "bucket already exists" error since
 		// `s3Svc.CreateBucket` does not error when the bucket exists
 		Config.Logger.Printf("Error creating S3 bucket '%s': %v", bucketName, err)
-		return err
+		return "", err
 	}
+	Config.Logger.Printf("INFO: Created S3 bucket '%s'", bucketName)
 
-	Config.Logger.Printf("Created S3 bucket '%s'", bucketName)
-	return nil
+	// set up KMS encryption on the bucket.
+	// the only way to check if the KMS key has already been created is to use an alias
+	kmsKeyAlias := fmt.Sprintf("alias/key-%s", bucketName)
+	kmsDescribeKeyOutput, err := kmsSvc.DescribeKey(&kms.DescribeKeyInput{
+		KeyId: aws.String(kmsKeyAlias),
+	})
+	var kmsKeyArn *string
+	if err == nil {
+		kmsKeyArn = kmsDescribeKeyOutput.KeyMetadata.Arn
+		Config.Logger.Printf("DEBUG: Existing KMS key: '%s' - '%s'", kmsKeyAlias, *kmsKeyArn)
+	} else {
+		// if the KMS key doesn't exist, create it
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == "NotFoundException" {
+			kmsCreateKeyOutput, err := kmsSvc.CreateKey(&kms.CreateKeyInput{
+				Tags: kmsTags,
+			})
+			if err != nil {
+				Config.Logger.Printf("Error creating KMS key: %v", err)
+				return "", err
+			}
+			kmsKeyArn = kmsCreateKeyOutput.KeyMetadata.Arn
+			Config.Logger.Printf("INFO: Created KMS key: '%s'", *kmsKeyArn)
+
+			_, err = kmsSvc.CreateAlias(&kms.CreateAliasInput{
+				AliasName:   &kmsKeyAlias,
+				TargetKeyId: kmsKeyArn,
+			})
+			if err != nil {
+				Config.Logger.Printf("Error creating KMS key alias: %v", err)
+				return "", err
+			}
```

Comment:
If for some reason we failed to create the Alias of the key, then are we going to handle it in anyway? In this case it looks like, we have created a bucket and a KMS key, where the bucket is not going to be KMS encrypted and the key is unusable.

Reply:
The key will be created but since the alias creation failed, the workspace launch fails.
Next time we launch, it will assume the key doesn't exist since it can't find it through the alias, so it will create a new key + alias.
So tldr it's not too bad - if the encryption fails, the workspace can't be launched at all